src/coreclr/jit/hwintrinsic.h

```diff
@@ -175,6 +176,18 @@ enum HWIntrinsicFlag : unsigned int
 
     // The intrinsic needs consecutive registers
     HW_Flag_NeedsConsecutiveRegisters = 0x4000,
+
+    // The intrinsic uses scalable registers
+    HW_Flag_Scalable = 0x8000,
+
+    // Returns Per-Element Mask
+    // the intrinsic returns a vector containing elements that are either "all bits set" or "all bits clear"
+    // this output can be used as a per-element mask
+    HW_Flag_ReturnsPerElementMask = 0x10000,
+
+    // The intrinsic uses a mask in arg1 to select elements present in the result
```

Comment:
`arg1`: Is it always be the case?

Reply:
Can we not just check for `TYP_MASK` to determine this?

Reply:
>arg1: Is it always be the case?

Yes, that's the sve convention. Result, then mask, then inputs.

>Can we not just check for TYP_MASK to determine this?

Ok, that sounds better. I can look and see how this would be done.

Reply:
> I imagine I'll have to write code that says if the actual type and expected type don't match, then somehow convert arg1 to the expected type.

Yes, basically.

Most intrinsics support masking optionally and so you'll have something similar to this https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/gentree.cpp#L19988-L20008. That is, you'll have some `bool GenTree::isSveEmbeddedMaskingCompatibleHWIntrinsic()` which likely looks up a flag in the `hwintrinsiclistarm64.h` table to see if that particular intrinsic supports embedded masking/predication.

There are then a handful of intrinsics which require masking. For example, SVE comparison intrinsics may always return a TYP_MASK, in which case you could either add a new entry to the table such as `HW_Flag_ReturnsSveMask` or explicitly handle it like xarch does here: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/hwintrinsicxarch.cpp#L3985-L3999

There are then a handful of intrinsics which require mask inputs and which aren't recognized via pattern matching. You would likewise add a flag or manually handle the few of them like this: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/hwintrinsicxarch.cpp#L3970-L3983

The insertion of the `ConvertVectorToMask` and `ConvertMaskToVector` intrinsics is important since the user may have passedin something that was of the incorrect type. For example, it might've been a mask of bytes, where we needed a mask of ints; or might've been an actual vector where we needed a mask and vice-versa. Likewise it ensures we don't need to check the type on every other intrinsic that does properly take a vector.

We then make this efficient in morph (see https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/morph.cpp#L10775-L10827) where we ensure that we aren't unnecessarily converting from mask to vector and back to mask, or vice versa. This allows things that take a mask to consume a produced mask directly and gives the optimal codegen expected in most scenarios.

Reply:
`Sve.TrueMask()` would be imported as `NI_SVE_ConvertMaskToVector(NI_SVE_TrueMask))`

`Sve.LoadVector(mask, address)` would then import as `NI_SVE_LoadVector(NI_SVE_ConvertVectorToMask(LCL_VAR), SomeNodeRepresentingAddress)`

Optimizations should kick in so we then get the following as the IR: `NI_SVE_LoadVector(NI_SVE_ConvertVectorToMask(NI_SVE_ConvertMaskToVector(NI_SVE_TrueMask))), SomeNodeRepresentingAddress)`

We would then fold that in morph to just `NI_SVE_LoadVector(NI_SVE_TrueMask, SomeNodeRepresentingAddress)` (and more ideally to just `NI_SVE_LoadVector(SomeNodeRepresentingAddress)` since the default `TrueMask` means include everything)

----

We are notably missing and need to add a bit which handles the case where we have `LCL_VAR TYP_SIMD = TYP_MASK` because that can currently block the ability to consume a mask directly if it's multi-use. We ideally would have it stored as `LCL_VAR TYP_MASK` instead (even if the use manually hoisted as a `Vector<T>` in C#/IL) and then have the things consume it as `ConvertMaskToVector(LCL_VAR)` if they actually needed a vector.

This shouldn't be overly complex to add, however, it's just not been done as of yet.

Reply:
> `Sve.TrueMask()` would be imported as `NI_SVE_ConvertMaskToVector(NI_SVE_TrueMask))`
> 
> `Sve.LoadVector(mask, address)` would then import as `NI_SVE_LoadVector(NI_SVE_ConvertVectorToMask(LCL_VAR), SomeNodeRepresentingAddress)`
> 
> Optimizations should kick in so we then get the following as the IR: `NI_SVE_LoadVector(NI_SVE_ConvertVectorToMask(NI_SVE_ConvertMaskToVector(NI_SVE_TrueMask))), SomeNodeRepresentingAddress)`
> 
> We would then fold that in morph to just `NI_SVE_LoadVector(NI_SVE_TrueMask, SomeNodeRepresentingAddress)` (and more ideally to just `NI_SVE_LoadVector(SomeNodeRepresentingAddress)` since the default `TrueMask` means include everything)
> 

Thanks! This is how I was understanding what to do. Hopefully we can drive all this with the use of enums in the hwinstrinsiclist.

Reply:
I'm not sure the lowering is enough to get rid of the masks. The following C#:

```csharp
Vector<UInt64> loadMask = Sve.CreateTrueMaskUInt64(SveMaskPattern.All);
var result = Sve.LoadVector(loadMask, (UInt64*)(_dataTable.inArray1Ptr));
```

Becomes:

```
***** BB01 [0000]
STMT00001 ( 0x00A[E-] ... ??? )
N004 (  7,  7) [000005] DA---------                         *  STORE_LCL_VAR simd16<System.Numerics.Vector`1> V01 loc0         
N003 (  3,  4) [000004] -----------                         \--*  HWINTRINSIC simd16 ulong ConvertMaskToVector
N002 (  2,  3) [000003] -----------                            \--*  HWINTRINSIC mask   ulong CreateTrueMaskUInt64
N001 (  1,  2) [000002] -------N---                               \--*  CNS_INT   int    31

***** BB01 [0000]
STMT00002 ( 0x012[E-] ... 0x023 )
N012 ( 33, 19) [000013] DACXGO-----                         *  STORE_LCL_VAR simd16<System.Numerics.Vector`1> V02 loc1         
N011 ( 29, 16) [000010] --CXGO--R--                         \--*  HWINTRINSIC simd16 ulong LoadVector
N010 (  5,  4) [000012] --------R--                            +--*  HWINTRINSIC mask   ulong ConvertVectorToMask
N009 (  1,  1) [000011] -----------                            |  +--*  HWINTRINSIC mask   ulong CreateTrueMaskAllUInt64
N008 (  3,  2) [000006] -----------                            |  \--*  LCL_VAR   simd16<System.Numerics.Vector`1> V01 loc0         
N007 ( 23, 11) [000009] --CXGO-----                            \--*  CALL      long   JIT.HardwareIntrinsics.Arm._Sve.LoadUnaryOpTest__SveLoadVector_ulong+DataTable:get_inArray1Ptr():ulong:this
N006 (  9,  8) [000037] ---X-O-N--- this in x0                    \--*  COMMA     byref 
N002 (  4,  3) [000033] ---X-O-----                                  +--*  NULLCHECK byte  
N001 (  3,  2) [000032] -----------                                  |  \--*  LCL_VAR   ref    V00 this         
N005 (  5,  5) [000036] -----O-----                                 Maybe  \--*  ADD       byref 
N003 (  3,  2) [000034] -----------                                     +--*  LCL_VAR   ref    V00 this         
N004 (  1,  2) [000035] -----------                                     \--*  CNS_INT   long   16 Fseq[_dataTable]

```

The `loadMask` is getting stored to memory as a lcl var, which makes sense. But that means it is going to be converted on every load/store for memory.

Maybe the lcl var can be eliminated because there is only a single use.

But what about when we have a bunch of masks being used multiple times? That's a lot of conversions.

Reply:
That was the comment around
> We are notably missing and need to add a bit which handles the case where we have LCL_VAR TYP_SIMD = TYP_MASK because that can currently block the ability to consume a mask directly if it's multi-use. We ideally would have it stored as LCL_VAR TYP_MASK instead (even if the use manually hoisted as a Vector<T> in C#/IL) and then have the things consume it as ConvertMaskToVector(LCL_VAR) if they actually needed a vector.
>
> This shouldn't be overly complex to add, however, it's just not been done as of yet.

Reply:
Right. That feels like it might touch quite a few files. Given the size of this PR, do you think it's worth keeping this PR as is, and then putting the `LCL_VAR TYP_MASK` in a follow on, along with the lowering code?

Reply:
> then putting the LCL_VAR TYP_MASK in a follow on

Yes, I think this would even be the preferred route given its not required and is its own isolated change really.

> along with the lowering code?

Which lowering code is this?

---------

In general I think its fine for this PR to be the basic plumbing of TYP_MASK support into the Arm64 side of the JIT. As long as `TrueMask` and `LoadVector` are minimally working as expected, I think we're golden and we can extend that to other operations and enable optimizations separately. That is exactly what we did for xarch to help with review and scoping.

Reply:
> Which lowering code is this?

I added some code do the remove the mask->vector->mask and vector->mask->vector conversions. But, nothing in this PR uses it because of the lcl var, so I decided not to push it.

Will mark this as ready now.

Reply:
> Will mark this as ready now.

... but not quite yet, as I need https://github.com/dotnet/runtime/pull/99049 to merge so I can remove it from this PR.